File: CucumberJVM/Shared/WIMP.App/java/com/wimp/app/services/OrderService.java
/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

package com.wimp.app.services;

import com.wimp.app.data.DataRepository;
import com.wimp.app.models.ContactDetails;
import com.wimp.app.models.DeliveryMethod;
import com.wimp.app.models.MenuItem;
import com.wimp.app.models.Order;
import com.wimp.app.models.OrderCollectionDetails;
import com.wimp.app.models.OrderStatus;
import com.wimp.app.models.PizzaItem;
import com.wimp.app.models.PizzaSize;
import com.wimp.app.models.ServiceResult;
import com.wimp.app.config.AppConfigurationProvider;
import com.wimp.app.restapi.PlaceOrderRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;

@Service
public class OrderService {
    private static final int DELIVERY_TIME_MINUTES = 45;
    private static final int MAX_LARGE_PIZZAS_PER_ORDER = 4;
    private static final Logger log = LoggerFactory.getLogger(OrderService.class);

    private final DataRepository repository;
    private final PromotionService promotionService;
    private final NotificationService notificationService;
    private final EmailService emailService;
    private final TimeService timeService;
    private final MenuService menuService;
    private final MessageService messageService;
    private final AppConfigurationProvider appConfigurationProvider;

    public OrderService(
        DataRepository repository,
        PromotionService promotionService,
        NotificationService notificationService,
        EmailService emailService,
        TimeService timeService,
        MenuService menuService,
        MessageService messageService,
        AppConfigurationProvider appConfigurationProvider
    ) {
        this.repository = repository;
        this.promotionService = promotionService;
        this.notificationService = notificationService;
        this.emailService = emailService;
        this.timeService = timeService;
        this.menuService = menuService;
        this.messageService = messageService;
        this.appConfigurationProvider = appConfigurationProvider;
    }

    public ServiceResult<Order> placeOrder(String customerName, PlaceOrderRequest placeOrderRequest) {
        var items = List.of(placeOrderRequest.items());
        int largePizzaCount = (int) items.stream().filter(item -> item.size() == PizzaSize.LARGE).count();

        LocalDateTime placingTime;
        LocalDateTime expectedDeliveryTime;
        if (appConfigurationProvider.backdoor().allowOverridingOrderTimes()) {
            placingTime = placeOrderRequest.placingTime() != null ? placeOrderRequest.placingTime() : timeService.getCurrentTime();
            expectedDeliveryTime = placeOrderRequest.expectedDeliveryTime() != null
                ? placeOrderRequest.expectedDeliveryTime()
                : placingTime.plusMinutes(DELIVERY_TIME_MINUTES);
        } else {
            placingTime = timeService.getCurrentTime();
            expectedDeliveryTime = placingTime.plusMinutes(DELIVERY_TIME_MINUTES);
        }

        if (largePizzaCount > MAX_LARGE_PIZZAS_PER_ORDER) {
            log.info("Rejecting order because of size limit");
            return ServiceResult.failure(messageService.getMessage("en-US", "cannot-deliver-too-many-large-pizzas", largePizzaCount));
        }

        BigDecimal price;
        if (appConfigurationProvider.backdoor().disableMenuIntegrityCheck()) {
            price = BigDecimal.valueOf(items.size()).multiply(BigDecimal.valueOf(12));
        } else {
            List<? extends MenuItem> menu = menuService.loadMenu();
            price = BigDecimal.ZERO;
            for (PizzaItem orderItem : items) {
                MenuItem menuItem = menu.stream()
                    .filter(mi -> mi.getName().equalsIgnoreCase(orderItem.name()))
                    .findFirst()
                    .orElse(null);
                if (menuItem == null) {
                    return ServiceResult.failure("Could not find pizza '" + orderItem.name() + "' on menu.");
                }
                price = price.add(menuItem.getPrice());
            }
        }

        Order order = new Order();
        order.setCustomerName(customerName);
        order.setCustomerEmail(placeOrderRequest.customerEmail());
        order.setItems(items);
        order.setDeliveryAddress(placeOrderRequest.deliveryAddress());
        order.setStatus(OrderStatus.PLACED);
        order.setPlacingTime(placingTime.toLocalTime());
        order.setExpectedDeliveryTime(expectedDeliveryTime);
        order.setPrice(price);
        order = repository.insertOrder(order);

        subscribeForDelayNotification(expectedDeliveryTime, order.getOrderNo(), customerName);
        if (appConfigurationProvider.simulation().orderRejection() && order.getPlacingTime().getSecond() == 30) {
            log.warn("Order {} has been rejected, because of a simulated error", order.getOrderNo());
            order = repository.updateOrder(order, OrderStatus.REJECTED, null, null, null);
        }
        log.info("Order {} placed", order.getOrderNo());
        return ServiceResult.success(order);
    }

    public Order getOrder(int orderNo) {
        return repository.getOrderByOrderNo(orderNo);
    }

    public ServiceResult<Order> cancelOrder(int orderNo, String customerName) {
        Order order = repository.getOrderByOrderNo(orderNo);
        if (order == null) {
            return ServiceResult.failure("Order " + orderNo + " not found.");
        }
        if (!order.getCustomerName().equalsIgnoreCase(customerName)) {
            return ServiceResult.failure("Order " + orderNo + " belongs to a different customer.");
        }
        if (order.getStatus() == OrderStatus.COMPLETED || order.getStatus() == OrderStatus.CANCELLED || order.getStatus() == OrderStatus.REJECTED) {
            return ServiceResult.failure("Order " + orderNo + " cannot be cancelled in its current status.");
        }

        order = repository.updateOrder(order, OrderStatus.CANCELLED, null, null, null);
        notificationService.sendCancellationNotification(customerName);
        return ServiceResult.success(order);
    }

    public ServiceResult<Order> setForCustomerCollection(int orderNo, String customerName) {
        Order order = repository.getOrderByOrderNo(orderNo);
        if (order == null) {
            return ServiceResult.failure("Order " + orderNo + " not found.");
        }
        if (!order.getCustomerName().equalsIgnoreCase(customerName)) {
            return ServiceResult.failure("Order " + orderNo + " belongs to a different customer.");
        }
        if (order.getDeliveryMethod() == DeliveryMethod.CUSTOMER_COLLECTION) {
            return ServiceResult.failure("Order " + orderNo + " is already set for customer-collection.");
        }
        if (order.getStatus() != OrderStatus.PLACED && order.getStatus() != OrderStatus.IN_PREPARATION) {
            return ServiceResult.failure("Order " + orderNo + " cannot be set to customer-collection in its current status.");
        }

        OrderCollectionDetails collectionDetails = new OrderCollectionDetails();
        collectionDetails.setOrderNo(orderNo);
        collectionDetails.setBoxesToBeCollected(order.getItems().size());
        collectionDetails.setContactDetailsConfirmationRequested(true);

        order = repository.updateOrder(order, null, null, DeliveryMethod.CUSTOMER_COLLECTION, collectionDetails);
        return ServiceResult.success(order);
    }

    public ServiceResult<Order> provideContactDetails(int orderNo, ContactDetails contactDetails) {
        Order order = repository.getOrderByOrderNo(orderNo);
        if (order == null) {
            return ServiceResult.failure("Order " + orderNo + " not found.");
        }

        String validationErrorMessage = validateContactDetails(contactDetails);
        if (validationErrorMessage == null && order.getDeliveryMethod() == DeliveryMethod.CUSTOMER_COLLECTION) {
            validationErrorMessage = validateCustomerCollectionContactDetails(contactDetails);
        }

        if (validationErrorMessage != null) {
            return ServiceResult.failure(validationErrorMessage);
        }

        order.setContactDetails(contactDetails);
        return ServiceResult.success(order);
    }

    private String validateCustomerCollectionContactDetails(ContactDetails contactDetails) {
        if ((contactDetails.getEmail() == null || contactDetails.getEmail().isBlank())
            && (contactDetails.getPhone() == null || contactDetails.getPhone().isBlank())) {
            return "For customer collection email or phone must be specified";
        }
        return null;
    }

    private String validateContactDetails(ContactDetails contactDetails) {
        if (contactDetails.getName() == null || contactDetails.getName().isBlank()) {
            return "Name not specified";
        }
        if (contactDetails.getEmail() != null && !contactDetails.getEmail().isBlank() && !isValidEmail(contactDetails.getEmail())) {
            return "Wrong email format";
        }
        if (contactDetails.getPhone() != null && !contactDetails.getPhone().isBlank() && !isValidPhone(contactDetails.getPhone())) {
            return "Wrong phone number format";
        }
        if ("US".equals(contactDetails.getCountry()) && (contactDetails.getState() == null || contactDetails.getState().isBlank() || "-".equals(contactDetails.getState()))) {
            return "For US country the state must be specified";
        }
        return null;
    }

    private boolean isValidPhone(String phone) {
        return phone.length() >= 6 && phone.chars().allMatch(Character::isDigit);
    }

    private boolean isValidEmail(String email) {
        return email.contains("@") && email.indexOf('@') > 0 && email.lastIndexOf('.') > email.indexOf('@');
    }

    public ServiceResult<Order> deliverOrder(int orderNo) {
        Order order = repository.getOrderByOrderNo(orderNo);
        if (order == null) {
            return ServiceResult.failure("Order " + orderNo + " not found.");
        }

        order = repository.updateOrder(order, OrderStatus.COMPLETED, null, null, null);

        boolean margheritaOrder = order.getItems().stream().anyMatch(item -> item.name().equalsIgnoreCase("Margherita"));
        if (margheritaOrder && promotionService.isPromotionActive("Margherita Friday")) {
            emailService.sendCoupon(order.getCustomerEmail(), "MARGHERITA25");
        }

        return ServiceResult.success(order);
    }

    public ServiceResult<Order> changeDeliveryAddress(int orderNo, String newAddress) {
        Order order = repository.getOrderByOrderNo(orderNo);
        if (order == null) {
            return ServiceResult.failure("Order " + orderNo + " not found.");
        }

        if (order.getStatus().compareTo(OrderStatus.WAITING_FOR_PICKUP) > 0) {
            return ServiceResult.failure("Delivery address can only be changed before the order is picked up.");
        }

        order = repository.updateOrder(order, null, newAddress, null, null);
        return ServiceResult.success(order);
    }

    public ServiceResult<Order> setWaitingForPickup(int orderNo) {
        Order order = repository.getOrderByOrderNo(orderNo);
        if (order == null) {
            return ServiceResult.failure("Order " + orderNo + " not found.");
        }
        if (order.getStatus() != OrderStatus.IN_PREPARATION) {
            return ServiceResult.failure("Order " + orderNo + " must be in preparation before it can wait for pickup.");
        }

        order = repository.updateOrder(order, OrderStatus.WAITING_FOR_PICKUP, null, null, null);
        return ServiceResult.success(order);
    }

    public Order takeNextOrder() {
        Order next = repository.getOrdersByStatus(OrderStatus.PLACED).stream().min(
                appConfigurationProvider.simulation().bug()
                    ? Comparator.comparing(Order::getPlacingTime).reversed()
                    : Comparator.comparing(Order::getPlacingTime))
            .orElse(null);
        if (next == null) {
            return null;
        }
        next = repository.updateOrder(next, OrderStatus.IN_PREPARATION, null, null, null);
        log.info("Taken next order: {}", next.getOrderNo());
        return next;
    }

    private void subscribeForDelayNotification(LocalDateTime expectedDeliveryTime, int orderNo, String customerName) {
        timeService.subscribeToTimeChange(time -> {
            if (time.isBefore(expectedDeliveryTime)) {
                return false;
            }

            Order order = repository.getOrderByOrderNo(orderNo);
            if (order != null && order.getStatus().compareTo(OrderStatus.COMPLETED) < 0) {
                notificationService.sendDelayNotification(customerName);
            }
            return true;
        });
    }
}

File: Reqnroll/Shared/WIMP.App/Services/OrderService.cs
/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

/*
 * NOTE: This application ("WIMP - Where Is My Pizza") is provided solely to
 * demonstrate the Behavior-Driven Development scenario automation patterns
 * described in the book "Effective Behavior-Driven Development" by
 * Gaspar Nagy and Seb Rose.
 *
 * It is NOT a complete or production-ready implementation. It deliberately
 * uses shortcuts and simplifications (e.g. authentication, data storage,
 * error handling, security) that are NOT suitable for a real application.
 * Do not use this code as a basis for production software.
 */

using System.Net.Mail;

using WIMP.App.Data;
using WIMP.App.Models;

namespace WIMP.App.Services;

public class OrderService(IDataRepository repository, PromotionService promotionService, NotificationService notificationService, EmailService emailService, ITimeService timeService, MenuService menuService, MessageService messageService, ILogger<OrderService> logger)
{
    private const int DeliveryTimeMinutes = 45;
    private const int MaxLargePizzasPerOrder = 4;

    public ServiceResult<Order> PlaceOrder(string customerName, string customerEmail, List<PizzaItem> items, string deliveryAddress, DateTimeOffset? testOnlyPlacingTime = null, DateTimeOffset? testOnlyExpectedDeliveryTime = null, bool testOnlyDisableMenuIntegrityCheck = true)
    {
        int largePizzaCount = items.Count(i => i.Size == PizzaSize.Large);
        var placingTime = timeService.GetCurrentTime();
#if DEBUG
        // Test-only API: for testing the placing time can be overridden
        placingTime = testOnlyPlacingTime ?? placingTime;
#endif
        var expectedDeliveryTime = placingTime.AddMinutes(DeliveryTimeMinutes);
#if DEBUG
        // Test-only API: for testing the expected delivery time can be overridden
        expectedDeliveryTime = testOnlyExpectedDeliveryTime ?? expectedDeliveryTime;
#endif
        if (largePizzaCount > MaxLargePizzasPerOrder)
        {
            logger.LogInformation("Rejecting order because of size limit");
            return ServiceResult<Order>.Failure(messageService.GetMessage("en-US", "cannot-deliver-too-many-large-pizzas", largePizzaCount));
        }

        decimal price;
        if (testOnlyDisableMenuIntegrityCheck)
        {
            price = items.Count * 12; // pricing simulation
        }
        else
        {
            var menu = menuService.LoadMenu();
            price = 0m;
            foreach (var orderItem in items)
            {
                var menuItem = menu.FirstOrDefault(mi => mi.Name == orderItem.Name);
                if (menuItem == null)
                {
                    return ServiceResult<Order>.Failure($"Could not find pizza '{orderItem.Name}' on menu.");
                }
                price += menuItem.Price;
            }
        }

        var order = repository.InsertOrder(new Order
        {
            CustomerName = customerName,
            CustomerEmail = customerEmail,
            Items = items,
            DeliveryAddress = deliveryAddress,
            Status = OrderStatus.Placed,
            PlacingTime = placingTime.TimeOfDay,
            Price = price
        });

        // scheduling to send a delay notification if the order is not delivered within 45 minutes
        SubscribeForDelayNotification(expectedDeliveryTime, order.OrderNo, customerName);

        logger.LogInformation("Order {OrderNo} placed", order.OrderNo);

#if SIMULATED_OREDER_REJECTION
        if (order.PlacingTime.Seconds == 30)
        {
            logger.LogWarning("Order {OrderNo} has been rejected, because of a simulated error", order.OrderNo);
            repository.UpdateOrder(order, status: OrderStatus.Rejected);
        }
#endif

        return ServiceResult<Order>.Success(order);
    }

    private void SubscribeForDelayNotification(DateTimeOffset expectedDeliveryTime, int orderNo, string customerName)
    {
        timeService.SubscribeToTimeChange(time =>
        {
            if (time >= expectedDeliveryTime)
            {
                var order = repository.GetOrderByOrderNo(orderNo);
                if (order is { Status: < OrderStatus.Completed })
                {
                    notificationService.SendDelayNotification(customerName);
                }
                return true;
            }
            return false;
        });
    }

    public Order? GetOrder(int orderNo)
    {
        return repository.GetOrderByOrderNo(orderNo);
    }

    public ServiceResult<Order> CancelOrder(int orderNo, string customerName)
    {
        var order = repository.GetOrderByOrderNo(orderNo);
        if (order is null)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} not found.");
        }

        if (!string.Equals(order.CustomerName, customerName, StringComparison.OrdinalIgnoreCase))
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} belongs to a different customer.");
        }

        if (order.Status is OrderStatus.Completed or OrderStatus.Cancelled or OrderStatus.Rejected)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} cannot be cancelled in its current status.");
        }

        order = repository.UpdateOrder(order, status: OrderStatus.Cancelled);
        notificationService.SendCancellationNotification(customerName);
        return ServiceResult<Order>.Success(order);
    }

    public ServiceResult<Order> SetForCustomerCollection(int orderNo, string customerName)
    {
        var order = repository.GetOrderByOrderNo(orderNo);
        if (order is null)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} not found.");
        }

        if (!string.Equals(order.CustomerName, customerName, StringComparison.OrdinalIgnoreCase))
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} belongs to a different customer.");
        }

        if (order.DeliveryMethod == DeliveryMethod.CustomerCollection)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} is already set for customer-collection.");
        }

        if (order.Status is not (OrderStatus.Placed or OrderStatus.InPreparation))
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} cannot be set to customer-collection in its current status.");
        }

        var collectionDetails = new OrderCollectionDetails
        {
            OrderNo = orderNo,
            BoxesToBeCollected = order.Items.Count,
            ContactDetailsConfirmationRequested = true
        };
        order = repository.UpdateOrder(order, deliveryMethod: DeliveryMethod.CustomerCollection, collectionDetails: collectionDetails);
        return ServiceResult<Order>.Success(order);
    }

    public ServiceResult<Order> ProvideContactDetails(int orderNo, ContactDetails contactDetails)
    {
        var order = repository.GetOrderByOrderNo(orderNo);
        if (order is null)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} not found.");
        }

        bool isValid = ValidateContactDetails(contactDetails, out string validationErrorMessage);
        if (isValid && order.DeliveryMethod == DeliveryMethod.CustomerCollection)
        {
            isValid = ValidateCustomerCollectionContactDetails(contactDetails, out validationErrorMessage);
        }

        if (!isValid)
        {
            return ServiceResult<Order>.Failure(validationErrorMessage);
        }

        order.ContactDetails = contactDetails;
        return ServiceResult<Order>.Success(order);
    }

    private bool ValidateCustomerCollectionContactDetails(ContactDetails contactDetails, out string errorMessage)
    {
        if (string.IsNullOrEmpty(contactDetails.Email) && string.IsNullOrEmpty(contactDetails.Phone))
        {
            errorMessage = "For customer collection email or phone must be specified";
            return false;
        }

        errorMessage = "";
        return true;
    }

    private bool ValidateContactDetails(ContactDetails contactDetails, out string errorMessage)
    {
        if (string.IsNullOrWhiteSpace(contactDetails.Name))
        {
            errorMessage = "Name not specified";
            return false;
        }

        if (!string.IsNullOrWhiteSpace(contactDetails.Email) && !IsValidEmail(contactDetails.Email))
        {
            errorMessage = "Wrong email format";
            return false;
        }

        if (!string.IsNullOrWhiteSpace(contactDetails.Phone) && !IsValidPhone(contactDetails.Phone))
        {
            errorMessage = "Wrong phone number format";
            return false;
        }

        if (contactDetails.Country == "US" && (string.IsNullOrWhiteSpace(contactDetails.State) || contactDetails.State == "none"))
        {
            errorMessage = "For US country the state must be specified";
            return false;
        }

        errorMessage = "";
        return true;
    }

    private bool IsValidPhone(string phone)
    {
        return phone.Length >= 6 && phone.All(char.IsAsciiDigit);
    }

    private static bool IsValidEmail(string email)
    {
        try
        {
            var mailAddress = new MailAddress(email);
            return mailAddress.Address == email;
        }
        catch (FormatException)
        {
            return false;
        }
    }

    public ServiceResult<Order> DeliverOrder(int orderNo)
    {
        var order = repository.GetOrderByOrderNo(orderNo);
        if (order is null)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} not found.");
        }

        order = repository.UpdateOrder(order, status: OrderStatus.Completed);

        if (order.Items.Any(i => string.Equals(i.Name, "Margherita", StringComparison.OrdinalIgnoreCase)) &&
            promotionService.IsPromotionActive("Margherita Friday"))
        {
            emailService.SendCoupon(order.CustomerEmail, "MARGHERITA25");
        }

        return ServiceResult<Order>.Success(order);
    }

    public ServiceResult<Order> ChangeDeliveryAddress(int orderNo, string newAddress)
    {
        var order = repository.GetOrderByOrderNo(orderNo);
        if (order is null)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} not found.");
        }

        if (order.Status > OrderStatus.WaitingForPickup)
        {
            return ServiceResult<Order>.Failure("Delivery address can only be changed before the order is picked up.");
        }

        order = repository.UpdateOrder(order, deliveryAddress: newAddress);
        return ServiceResult<Order>.Success(order);
    }

    public ServiceResult<Order> SetWaitingForPickup(int orderNo)
    {
        var order = repository.GetOrderByOrderNo(orderNo);
        if (order is null)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} not found.");
        }

        if (order.Status != OrderStatus.InPreparation)
        {
            return ServiceResult<Order>.Failure($"Order {orderNo} must be in preparation before it can wait for pickup.");
        }

        order = repository.UpdateOrder(order, status: OrderStatus.WaitingForPickup);
        return ServiceResult<Order>.Success(order);
    }

    public Order? TakeNextOrder()
    {
        var next = repository.GetOrdersByStatus(OrderStatus.Placed)
#if SIMULATED_BUG
            .OrderByDescending(o => o.PlacingTime)
#else
            .OrderBy(o => o.PlacingTime)
#endif
            .FirstOrDefault();

        if (next is null)
        {
            return null;
        }

        next = repository.UpdateOrder(next, status: OrderStatus.InPreparation);
        logger.LogInformation("Taken next order: {OrderNo}", next.OrderNo);
        return next;
    }
}
